unumpy/_multimethods.py

```diff
@@ -322,6 +322,17 @@ def accumulate(self, a, axis=0, dtype=None, out=None):
 trunc = ufunc("trunc", 1, 1)
 
 
+@create_numpy(_dtype_argreplacer)
+def empty(shape, dtype=float, order='C'):
+    return (mark_dtype(dtype),)
+
+
+@create_numpy(_dtype_argreplacer)
```

Comment:
You will need to create a new argument replacer, `_self_dtype_argreplacer`.

Reply:
A `default=` needs to be added that implements `empty_like` in terms of `empty`.
```suggestion
@create_numpy(_dtype_argreplacer, default=lambda prototype, dtype=None, order="K", subok=True, shape=None: empty(shape(prototype), prototype.dtype))
```

Reply:
> A `default=` needs to be added that implements `empty_like` in terms of `empty`.

I have some questions about this:
- For example, if the kwarg `shape` is passed it overrides the shape of `prototype` and this behaviour is not foreseen in this default implementation. It seems that these implementations can be arbitrarily simple or complex. With that said, is there a general rule of thumb on writing these?
- It seems that not all multimethods need a default implementation so in which scenarios should they be considered?

Thank you!

Reply:
>     * For example, if the kwarg `shape` is passed it overrides the shape of `prototype` and this behaviour is not foreseen in this default implementation. It seems that these implementations can be arbitrarily simple or complex. With that said, is there a general rule of thumb on writing these?

You're right -- My initial implementation was incorrect, good catch. I'd say it depends: I'd say if there's an implementation that's under 20-30 LOC and reasonably fast (it contains no for loops and so on), it'd be good to write it.

One exception to this was `block`, as it was a fairly important function.

Reply:
Here is a possible implementation:

```python
def _empty_like_default(prototype, dtype=None, order="K", subok=True, shape=None):
    if shape:
        return empty(shape, dtype=prototype.dtype)

    return empty(shape(prototype), dtype=prototype.dtype)
```

**Note:** This does not capture all possible behaviour as I'm not sure how to get the `order` of `prototype` in order to pass it to the new array. Maybe through `ndarray.flags`?

Reply:
> You will need to create a new argument replacer, `_self_dtype_argreplacer`.

Not exactly sure how to do this but I tried to bundle `_self_argreplacer` and `_dtype_argreplacer` together in some way. Here is the result:

```python
def _self_dtype_argreplacer(args, kwargs, dispatchables):
    def replacer(a, *args, dtype=None, **kwargs):
        out_kw = kwargs.copy()
        out_kw["dtype"] = dispatchables[1]

        return (dispatchables[0],) + args, out_kw

    return replacer(*args, **kwargs)
```

It seems to work although I haven't fully tested this.

Reply:
> **Note:** This does not capture all possible behaviour as I'm not sure how to get the `order` of `prototype` in order to pass it to the new array. Maybe through `ndarray.flags`?

I think we should explicitly check for `shape is None` as this will also match empty tuples, which is a bug.



> Not exactly sure how to do this but I tried to bundle `_self_argreplacer` and `_dtype_argreplacer` together in some way. Here is the result:

That looks about right.

Reply:
We should also check for `dtype` the same way as `shape` in the default implementation.